sql/schemachanger: track column and index dependencies on triggers

spilchen opened the pull request:
Enhance the schema changer to record dependencies between triggers and the columns or indexes they reference. Previously, these dependencies were not tracked, allowing operations such as dropping a column or index used by a trigger to proceed without error.

The TriggerDeps descriptor element, which previously tracked dependencies as simple table references, now records column and index IDs per table. A new field, `UsesRelations`, captures these finer-grained dependencies and replaces the now-deprecated `UsesRelationIDs`.

To reduce the scope of this change, dependency tracking is currently omitted for triggers that reference their own table. This limitation will be addressed in a future update. I figured this was fine so it's no worse than what we have now.

Informs: #142370

Epic: none
Release note (bug fix): Prevent dropping columns or indexes that are still referenced by triggers. Previously, these operations could succeed silently, potentially breaking trigger functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146683)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
this looks good so far! i have pretty small comments. i agree with merging this PR, and finishing up the self-reference tracking in the next PR so we can fully resolve https://github.com/cockroachdb/cockroach/issues/142370

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-:-OQQ9PGyB80c1n0YsLkN:b-auzijq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ6Wos4D92QIOePxZA:-OQQ6Wos4D92QIOePxZB:by6ayi4):*
nit: let's add a bug fix release note for this.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 900 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ7Xw51UFww0U2MUc-:-OQQ7Xw51UFww0U2MUc0:b-qika16) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/sql/schemachanger/scdecomp/decomp.go#L900)):*
> ```Go
> 			panic(err)
> 		}
> 		_ = toDesc.ForeachDependedOnBy(func(dep *descpb.TableDescriptor_Reference) error {
> ```

although there's no error returned currently, let's still check for one and handle it so this code is resilient for the future in case we do add an error later.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/triggers` line 4378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ8wCm443Xv_54dbag:-OQQ8wCm443Xv_54dbah:b8o3w2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/ccl/logictestccl/testdata/logic_test/triggers#L4378)):*
> ```
> DROP FUNCTION f1 CASCADE;
> 
> statement error pq: cannot drop type "e" because other objects \(\[test.public.t1\]\) still depend on it
> ```

is it possible to define a trigger that references this type, even if there is no table that uses the type? we would still want to block dropping the type in that case.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go` line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ8PNFBGN2Y18t9krk:-OQQ8PNFBGN2Y18t9krl:bf2hi22) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go#L367)):*
> ```Go
> }
> 
> // maybeDropDependentViews attempts to drop all views that depend on the index
> ```

nit: should say `maybeDropTriggers` and adjust the wording to stop mentioning views.
___
*[`pkg/sql/schemachanger/scpb/migration.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ7zTL6L9YTC8oY5bM:-OQQ7zTL6L9YTC8oY5bN:b2orxqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/sql/schemachanger/scpb/migration.go#L47)):*
> ```Go
> 
> 	// In TriggerDeps, map the deprecated UsesRelationIDs to a
> 	// TriggerDeps_RelationReference.
> ```

i'm forgetting some details on how this migration works -- specifically i'm wondering if we need a version gate in the other places of the code that use `UsesRelations` to make sure we are on a finalized cluster version. also, does this mean we won't be able to backport?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: track column and index dependencies on triggers

Enhance the schema changer to record dependencies between triggers and the
columns or indexes they reference. Previously, these dependencies were not
tracked, allowing operations such as dropping a column or index used by a
trigger to proceed without error.

The TriggerDeps descriptor element, which previously tracked dependencies
as simple table references, now records column and index IDs per table. A new
field, UsesRelations, captures these finer-grained dependencies and replaces
the now-deprecated UsesRelationIDs.

To reduce the scope of this change, dependency tracking is currently omitted
for triggers that reference their own table. This limitation will be addressed
in a future update. The current behavior remains consistent with the previous
state.

Informs: #142370
Release note (bug fix): Prevent dropping columns or indexes that are
still referenced by triggers. Previously, these operations could succeed
silently, potentially breaking trigger functionality.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 900 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ7Xw51UFww0U2MUc-:-OQhd_uK2iSm5_0_Td4G:b-ip0uc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/sql/schemachanger/scdecomp/decomp.go#L900)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

although there's no error returned currently, let's still check for one and handle it so this code is resilient for the future in case we do add an error later.
</blockquote></details>

Done. This copied a pattern we already had in this file, so I changed a few other places that did this too.
___
*[`pkg/sql/schemachanger/scpb/migration.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ7zTL6L9YTC8oY5bM:-OQhzL0n4E3zC0mVq-aY:b17d4b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/sql/schemachanger/scpb/migration.go#L47)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm forgetting some details on how this migration works -- specifically i'm wondering if we need a version gate in the other places of the code that use `UsesRelations` to make sure we are on a finalized cluster version. also, does this mean we won't be able to backport?
</blockquote></details>

I think this migration code is here for a restore, when the backup was taken and the job existed but not yet run. It runs the migration code to update DSC elements to fill in the new UsesRelations field.
    
The CREATE TRIGGER DDL is pretty simple from a DSC perspective. It has the minimum number of stages. There is no long running backfill or validation step. And since it's updates to metadata only, it won't be planned across many nodes in the cluster (aside from the waitForOneVersion logic). None of the dep rules rely on the deprecated UsesRelationIDs field. So, given all of this, I think it doesn't need version gates and is fine to backport. It ends up being all or nothing. If it's planned on a node with the new code changes, it will use the new code. Otherwise, it will fallback to the old code.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/triggers` line 4378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ8wCm443Xv_54dbag:-OQhvjsw9rkTGgQg6r2q:b-1ylned) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/ccl/logictestccl/testdata/logic_test/triggers#L4378)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is it possible to define a trigger that references this type, even if there is no table that uses the type? we would still want to block dropping the type in that case.
</blockquote></details>

Yes, that's what we are doing in this test. Notice, we only use the type `e` by projecting it from a SELECT statement, like this:
`SELECT *, 'a'::e FROM t1@i1 WHERE j=32;`


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-:-OQiR2Wj2nJEYffKZ2Qe:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 900 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ7Xw51UFww0U2MUc-:-OQiQoVOBNw6sTs0Lpaq:b7ajauf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/sql/schemachanger/scdecomp/decomp.go#L900)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Done. This copied a pattern we already had in this file, so I changed a few other places that did this too.
</blockquote></details>

thanks, nice catch!
___
*[`pkg/ccl/logictestccl/testdata/logic_test/triggers` line 4378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146683#-OQQ8wCm443Xv_54dbag:-OQiR-0O-fqWsPCMJThw:bwamq24) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e8d0cbd3160444a192f3cf0135469f69adb812a/pkg/ccl/logictestccl/testdata/logic_test/triggers#L4378)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Yes, that's what we are doing in this test. Notice, we only use the type `e` by projecting it from a SELECT statement, like this:
`SELECT *, 'a'::e FROM t1@i1 WHERE j=32;`
</blockquote></details>

thanks for explaining, i had mis-interpreted the error message.


<!-- Sent from Reviewable.io -->
